src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -346,6 +417,168 @@ private BoundExpression RewriteStringConcatenationManyExprs(SyntaxNode syntax, I
+                        readOnlySpanCtorRefParamChar,
+                        preparedArgs.ToImmutableAndFree());
+
+                return true;
+            }
+
+            result = null;
+            return false;
+
+            static bool tryGetNeededToSpanMembers(
+                LocalRewriter self,
+                SyntaxNode syntax,
+                bool needsImplicitConversionFromStringToSpan,
+                NamedTypeSymbol charType,
+                [NotNullWhen(true)] out MethodSymbol? readOnlySpanCtorRefParamChar,
+                out MethodSymbol? stringImplicitConversionToReadOnlySpan)
+            {
+                readOnlySpanCtorRefParamChar = null;
+                stringImplicitConversionToReadOnlySpan = null;
+
+                if (self.TryGetWellKnownTypeMember(syntax, WellKnownMember.System_ReadOnlySpan_T__ctor_Reference, out MethodSymbol? readOnlySpanCtorRefParamGeneric, isOptional: true))
+                {
+                    var readOnlySpanOfChar = readOnlySpanCtorRefParamGeneric.ContainingType.Construct(charType);
+                    readOnlySpanCtorRefParamChar = readOnlySpanCtorRefParamGeneric.AsMember(readOnlySpanOfChar);
+                }
+                else
+                {
+                    return false;
+                }
+
+                if (needsImplicitConversionFromStringToSpan)
+                {
+                    return self.TryGetWellKnownTypeMember(syntax, WellKnownMember.System_String__op_Implicit_ToReadOnlySpanOfChar, out stringImplicitConversionToReadOnlySpan, isOptional: true);
+                }
+
+                return true;
+            }
+
+            static BoundExpression rewriteStringConcatenationWithSpanBasedConcat(
+                SyntaxNode syntax,
+                SyntheticBoundNodeFactory factory,
+                MethodSymbol spanConcat,
+                MethodSymbol? stringImplicitConversionToReadOnlySpan,
+                MethodSymbol readOnlySpanCtorRefParamChar,
+                ImmutableArray<BoundExpression> args)
+            {
+                var preparedArgsBuilder = ArrayBuilder<BoundExpression>.GetInstance(capacity: args.Length);
+                var localsBuilder = ArrayBuilder<LocalSymbol>.GetInstance();
+
+                foreach (var arg in args)
+                {
+                    Debug.Assert(arg.Type is not null);
+
+                    if (arg.Type.SpecialType == SpecialType.System_Char)
+                    {
+                        var temp = factory.StoreToTemp(arg, out var tempAssignment);
+                        localsBuilder.Add(temp.LocalSymbol);
+
+                        var wrappedChar = new BoundObjectCreationExpression(
+                            arg.Syntax,
+                            readOnlySpanCtorRefParamChar,
+                            [temp],
+                            argumentNamesOpt: default,
+                            argumentRefKindsOpt: [RefKindExtensions.StrictIn],
```

Comment:
@AlekseyTs When I tested my code with custom corlib I also found a crash: if corlib declares ROS constructor as just `ref` instead of `ref readonly` we crash because well-known members only contain info about whether the argument is passed by reference, but cannot specify what kind of reference it should be. Thus we crash here because I've hardcoed `RefKindExtensions.StrictIn`. Does this thing need fixing (I guess yes)? What fix should we apply: gracefully fallback to just `ref` as ref kind or avoid span-based concat in such case entirely?

Reply:
> What fix should we apply

I think we should simply check parameter's ```RefKind``` here and use ```RefKind.Ref``` if it is a plain ```RefKind.Ref```. if it is `in` or `ref readonly`, we should use ```RefKindExtensions.StrictIn``` or ```RefKind.In```. If I remember correctly, ```RefKindExtensions.StrictIn``` prevents async rewrite from spilling the argument by value. But in our case that should be fine, no one else is using the local and we do not plan to change its value. 

BTW, it is good this issue came up, we actually need to test whether async rewriter is going to allow and properly handle usage of span types in this manner. What we need to test, is a scenario when the next operand of concatenation is an `await` expression. This should force the spilling of the span instance that we create here. Similar scenario should be tested with span created from a string operand. It is quite possible we will have to "teach" ```SpillSequenceSpiller``` to spill these spans properly.

It might be too hard to test async scenarios against custom core library
But we don't really need that in order to have a custom ```ReadOnlySpan`` type. It should be possible to simply defining it in the same compilation. We should try testing all `ref` flavors for the parameter.

Reply:
>It might be too hard to test async scenarios against custom core library
>But we don't really need that in order to have a custom ```ReadOnlySpan``` type.

Scratch that, I forgot about the String.Concat helper that is supposed to use the same ```ReadOnlySpan``` type. Let's start with async scenario for the official core library then.

Reply:
> Let's start with async scenario for the official core library then

Can you elaborate on this please? `async` will generate a state machine, which I am not really familiar with. What things are we interested in to verify? Some specific method of the machine, maybe fields declarations (to ensure span isn't leaking to the heap)?

Reply:
> Can you elaborate on this please

Simply put, use concatenation in an `async` method where each operand is an `await` expression. Verify runtime behavior and IL for the state machine method.

Reply:
Variations can be made in terms of what operands use `await`, but having it only on the first operand is probably not very interesting for the test purposes.

Reply:
> I think we should simply check parameter's `RefKind` here and use `RefKind.Ref` if it is a plain `RefKind.Ref`. if it is `in` or `ref readonly`, we should use `RefKindExtensions.StrictIn` or `RefKind.In`

There is also `RefKind.Out`, which is technically possible there as well. For now I made in unacceptable ref kind, so if ROS reference constructor is by `out` we fallback to string-based concat. LMK if you think the behavior should be different

Reply:
> LMK if you think the behavior should be different

I think this is a reasonable way to deal with the situation.

Reply:
> Simply put, use concatenation in an `async` method where each operand is an `await` expression. Verify runtime behavior and IL for the state machine method.

Here is the source I tried:
```cs
using System;
using System.Threading.Tasks;

public class Test
{
    static async Task Main()
    {
        Console.Write(await M());
    }

    static async Task<string> M()
    {
        return (await GetStringTask()) + (await GetCharTask());
    }

    static Task<string> GetStringTask()
        => Task.FromResult("s");

    static Task<char> GetCharTask()
        => Task.FromResult('c');
}
```
`CompileAndVerify` failed with `error CS4007: 'await' cannot be used in expression, containing 'System.ReadOnlySpan<char>' type`. So what should we do? Ban span-based optimization in `async` context (if so, how to detect that we are in `async` context) or somehow whitelist our optimization to not trigger the error (how can we ensure ROS isn't escaping to the heap then?)

Reply:
Yes this is the kind of test scenario that I had in mind. I would like to suggest converting ```GetStringTask```/```GetCharTask``` to async methods as well and doing ```await Task.Yield();``` before returning from them. That will ensure we will be testing "real" async behavior at runtime. Also, I think we need to have a test for at least one more scenario, a scenario with `awat` in the operand after the `char` operand.

> So what should we do? Ban span-based optimization in async context (if so, how to detect that we are in async context) or somehow whitelist our optimization to not trigger the error (how can we ensure ROS isn't escaping to the heap then?)

No. I think we will be able to avoid an attempt to spill the ```ReadOnlySpan``` instances that we create for concatenation. We are already doing something similar for other features. Look for the following comment in ```SpillSequenceSpiller.Spill``` method:

> // If the method is known to have no observable side-effects, we can spill its receiver and arguments,
> // and call it later, after some other (unrelated) side-effects are evaluated.

We need to add handling for an additional ```BoundCall``` shape, and for ```BoundObjectCreationExpression``` that we create here in that function.